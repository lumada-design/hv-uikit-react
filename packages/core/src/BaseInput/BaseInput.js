import React, { useContext } from "react";
import PropTypes from "prop-types";
import clsx from "clsx";
import { Input, withStyles } from "@material-ui/core";
import { refType } from "../utils";
import styles from "./styles";

import {
  HvFormElementContext,
  HvFormElementDescriptorsContext,
  buildFormElementPropsFromContext,
  buildAriaPropsFromContext,
} from "../Forms/FormElement";

/**
 * An Input component that only posses the most basic functionalities.
 * It should be used alongside the other form elements to construct a proper accessible form.
 */
const HvBaseInput = (props) => {
  const {
    classes,
    className = "",

    id,
    name,

    value,
    defaultValue,

    required,
    readOnly,
    disabled,

    onChange,

    type = "text",

    placeholder,

    multiline = false,
    resizable = false,
    invalid = false,

    inputRef,
    inputProps = {},

    ...others
  } = props;

  const formElementContext = useContext(HvFormElementContext);
  const formElementProps = buildFormElementPropsFromContext(props, formElementContext);

  const localInvalid = invalid || formElementProps.status === "invalid";

  const formElementDescriptorsContext = useContext(HvFormElementDescriptorsContext);
  const ariaProps = buildAriaPropsFromContext(
    inputProps,
    formElementDescriptorsContext,
    localInvalid,
    id
  );

  const onChangeHandler = (event) => {
    onChange?.(event, event.target.value);
  };

  return (
    <div
      className={clsx(classes.root, className, {
        [classes.disabled]: formElementProps.disabled,
        [classes.invalid]: localInvalid,
        [classes.resizable]: multiline && resizable,
        [classes.readOnly]: readOnly,
      })}
    >
      <Input
        id={id}
        name={formElementProps.name}
        value={value}
        defaultValue={defaultValue}
        type={type}
        placeholder={placeholder}
        readOnly={formElementProps.readOnly}
        disabled={formElementProps.disabled}
        onChange={onChangeHandler}
        classes={{
          root: classes.inputRoot,
          focused: classes.inputRootFocused,
          disabled: classes.inputRootDisabled,
          multiline: classes.inputRootMultiline,
          input: clsx(classes.input, {
            [classes.inputResizable]: !formElementProps.disabled && resizable,
            [classes.inputDisabled]: disabled,
          }),
        }}
        inputProps={{
          // avoid the required attribute at the root node
          required: formElementProps.required,
          ...inputProps,
          ...ariaProps,
        }}
        inputRef={inputRef}
        multiline={multiline}
        rows={10}
        {...others}
      />
      {!multiline && <div role="presentation" className={classes.inputBorderContainer} />}
    </div>
  );
};

HvBaseInput.formElementType = "input";

HvBaseInput.propTypes = {
  /**
   * Class names to be applied.
   */
  className: PropTypes.string,
  /**
   * A Jss Object used to override or extend the component styles applied.
   */
  classes: PropTypes.shape({
    /**
     * Styles applied to the root container of the input.
     */
    root: PropTypes.string,
    /**
     * Styles applied to the root container of the input when it is disabled.
     */
    disabled: PropTypes.string,
    /**
     * Styles applied to the root container of the input when it is invalid.
     */
    invalid: PropTypes.string,
    /**
     * Styles applied to the root container of the input when it is resizable.
     */
    resizable: PropTypes.string,

    /**
     * Styles applied to input root which is the input that encloses all the other elements.
     */
    inputRoot: PropTypes.string,
    /**
     * Styles applied to input root when it is focused.
     */
    inputRootFocused: PropTypes.string,
    /**
     * Styles applied to input html element when it is disabled.
     */
    inputRootDisabled: PropTypes.string,
    /**
     * Styles applied to input html element when it is multiline mode.
     */
    inputRootMultiline: PropTypes.string,

    /**
     * Styles applied to input html element.
     */
    input: PropTypes.string,
    /**
     * Styles applied to input html element when is disabled.
     */
    inputDisabled: PropTypes.string,
    /**
     * Styles applied to input html element when it is resizable.
     */
    inputResizable: PropTypes.string,

    /**
     * Styles applied to the container of the border element.
     */
    inputBorderContainer: PropTypes.string,
    /**
     * Styles applied to the container of the border element, when in read only mode.
     */
    readOnly: PropTypes.string,
  }).isRequired,

  /**
   * Id to be applied to the root node.
   */
  id: PropTypes.string,
  /**
   * The input name.
   */
  name: PropTypes.string,

  /**
   * The value of the input, when controlled.
   */
  value: PropTypes.string,
  /**
   * The initial value of the input, when uncontrolled.
   */
  defaultValue: PropTypes.string,

  /**
   * If `true` the input is disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Indicates that the input is not editable.
   */
  readOnly: PropTypes.bool,
  /**
   * If true, the input element will be required.
   */
  required: PropTypes.bool,

  /**
   * The function that will be executed onChange, allows modification of the input,
   * it receives the value. If a new value should be presented it must returned it.
   */
  onChange: PropTypes.func,

  /**
   * The input type.
   */
  type: PropTypes.string,

  /**
   * Label inside the input used to help user.
   */
  placeholder: PropTypes.string,

  /**
   * If true, a textarea element will be rendered.
   */
  multiline: PropTypes.bool,
  /**
   * If true and multiline is also true the textarea element will be resizable.
   */
  resizable: PropTypes.bool,
  /**
   * Denotes if the input is in an invalid state.
   */
  invalid: PropTypes.bool,

  /**
   * Attributes applied to the input element.
   */
  inputProps: PropTypes.instanceOf(Object),
  /**
   * Allows passing a ref to the underlying input
   */
  inputRef: refType,
};

export default withStyles(styles, { name: "HvBaseInput" })(HvBaseInput);
