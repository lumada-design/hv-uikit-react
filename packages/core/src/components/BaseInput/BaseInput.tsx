import { RefObject, useContext } from "react";
import {
  InputBaseComponentProps as MuiInputBaseComponentProps,
  InputProps as MuiInputProps,
} from "@mui/material";
import { HvBaseProps } from "../../types";
import {
  StyledRoot,
  StyledInputBorderContainer,
  StyledInput,
} from "./BaseInput.styles";
import {
  HvFormElementContext,
  HvFormElementDescriptorsContext,
  buildFormElementPropsFromContext,
  buildAriaPropsFromContext,
} from "../Forms/FormElement";
import clsx from "clsx";

export type HvBaseInputProps = Omit<MuiInputProps, "onChange"> &
  HvBaseProps<HTMLDivElement, { onChange }> & {
    /** The input name. */
    name?: string;
    /** The value of the input, when controlled. */
    value?: string;
    /** The initial value of the input, when uncontrolled. */
    defaultValue?: string;
    /** If `true` the input is disabled. */
    disabled?: boolean;
    /** Indicates that the input is not editable. */
    readOnly?: boolean;
    /** If true, the input element will be required. */
    required?: boolean;
    /** The function that will be executed onChange, allows modification of the input,
     * it receives the value. If a new value should be presented it must returned it. */
    onChange?: (
      event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
      value: string
    ) => void;
    /** The input type. */
    type?: string;
    /** Label inside the input used to help user. */
    placeholder?: string;
    /** If true, a textarea element will be rendered. */
    multiline?: boolean;
    /** If true and multiline is also true the textarea element will be resizable. */
    resizable?: boolean;
    /** Denotes if the input is in an invalid state. */
    invalid?: boolean;
    /** Attributes applied to the input element. */
    inputProps?: MuiInputBaseComponentProps;
    /** Allows passing a ref to the underlying input */
    inputRef?: RefObject<HTMLElement>;
    /** A Jss Object used to override or extend the styles applied to the empty state component. */
    classes?: {
      /** Styles applied to the root container of the input. */
      root?: string;
      /** Styles applied to the root container of the input when it is disabled. */
      disabled?: string;
      /** Styles applied to the root container of the input when it is invalid. */
      invalid?: string;
      /** Styles applied to the root container of the input when it is resizable. */
      resizable?: string;
      /** Styles applied to input root which is the input that encloses all the other elements. */
      inputRoot?: string;
      /** Styles applied to input root when it is focused. */
      inputRootFocused?: string;
      /** Styles applied to input html element when it is disabled. */
      inputRootDisabled?: string;
      /** Styles applied to input html element when it is multiline mode. */
      inputRootMultiline?: string;
      /** Styles applied to input html element. */
      input?: string;
      /** Styles applied to input html element when is disabled. */
      inputDisabled?: string;
      /** Styles applied to input html element when it is resizable. */
      inputResizable?: string;
      /** Styles applied to the container of the border element. */
      inputBorderContainer?: string;
      /** Styles applied to the container of the border element, when in read only mode. */
      readOnly?: string;
    };
  };

/**
 * An Input component that only posses the most basic functionalities.
 * It should be used alongside the other form elements to construct a proper accessible form.
 */
export const HvBaseInput = ({
  classes,
  className = "",
  id,
  name,
  value,
  defaultValue,
  required,
  readOnly,
  disabled,
  onChange,
  type = "text",
  placeholder,
  multiline = false,
  resizable = false,
  invalid = false,
  inputRef,
  inputProps = {},
  ...others
}: HvBaseInputProps) => {
  const formElementContext = useContext(HvFormElementContext);
  const formElementProps = buildFormElementPropsFromContext(
    name,
    disabled,
    readOnly,
    required,
    status,
    formElementContext
  );

  const localInvalid = invalid || formElementProps.status === "invalid";

  const formElementDescriptorsContext = useContext(
    HvFormElementDescriptorsContext
  );
  const ariaProps = buildAriaPropsFromContext(
    inputProps,
    formElementDescriptorsContext,
    localInvalid,
    id
  );

  const onChangeHandler = (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    onChange?.(event, event.target.value);
  };

  return (
    <StyledRoot
      className={clsx(
        "root",
        classes?.root,
        className,
        formElementProps.disabled && classes?.disabled,
        localInvalid && classes?.invalid,
        multiline && resizable && classes?.resizable,
        readOnly && classes?.readOnly
      )}
      $disabled={formElementProps.disabled}
      $invalid={localInvalid}
      $resizable={multiline && resizable}
      $readOnly={!!readOnly}
    >
      <StyledInput
        id={id}
        name={formElementProps.name}
        value={value}
        defaultValue={defaultValue}
        type={type}
        placeholder={placeholder}
        readOnly={!!formElementProps.readOnly}
        disabled={formElementProps.disabled}
        onChange={onChangeHandler}
        className={clsx(
          localInvalid && "inputRootInvalid",
          readOnly && "inputRootReadOnly"
        )}
        classes={{
          root: clsx("inputRoot", classes?.inputRoot),
          focused: clsx("inputRootFocused", classes?.inputRootFocused),
          disabled: clsx("inputRootDisabled", classes?.inputRootDisabled),
          multiline: clsx("inputRootMultiline", classes?.inputRootMultiline),
          input: clsx(
            "input",
            classes?.input,
            !formElementProps.disabled &&
              resizable &&
              clsx("inputResizable", classes?.inputResizable),
            disabled && clsx("inputDisabled", classes?.inputDisabled),
            readOnly && "readOnly" && classes?.readOnly
          ),
        }}
        inputProps={{
          // Avoid the required attribute at the root node
          required: formElementProps.required,
          ...inputProps,
          ...ariaProps,
        }}
        inputRef={inputRef}
        multiline={multiline}
        rows={10}
        $resizable={!formElementProps.disabled && resizable}
        $disabled={!!disabled}
        $readOnly={!!readOnly}
        $invalid={localInvalid}
        {...others}
      />
      {!multiline && (
        <StyledInputBorderContainer
          role="presentation"
          className={clsx(
            "inputBorderContainer",
            classes?.inputBorderContainer
          )}
        />
      )}
    </StyledRoot>
  );
};
