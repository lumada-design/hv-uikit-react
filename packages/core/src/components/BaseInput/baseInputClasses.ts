import { getClasses } from "utils";

export type HvBaseInputClasses = {
  /** Styles applied to the root container of the input. */
  root?: string;
  /** Styles applied to the root container of the input when it is disabled. */
  disabled?: string;
  /** Styles applied to the root container of the input when it is invalid. */
  invalid?: string;
  /** Styles applied to the root container of the input when it is resizable. */
  resizable?: string;
  /** Styles applied to input root which is the input that encloses all the other elements. */
  inputRoot?: string;
  /** Styles applied to input root when it is focused. */
  inputRootFocused?: string;
  /** Styles applied to input root element when it is disabled. */
  inputRootDisabled?: string;
  /** Styles applied to input root element when it is multiline mode. */
  inputRootMultiline?: string;
  /** Styles applied to input root element when it is invalid. */
  inputRootInvalid?: string;
  /** Styles applied to input root element when it is read only. */
  inputRootReadOnly?: string;
  /** Styles applied to input html element. */
  input?: string;
  /** Styles applied to input html element when is disabled. */
  inputDisabled?: string;
  /** Styles applied to input html element when it is resizable. */
  inputResizable?: string;
  /** Styles applied to input html element when it is read only. */
  inputReadOnly?: string;
  /** Styles applied to the container of the border element. */
  inputBorderContainer?: string;
  /** Styles applied to the container of the border element, when in read only mode. */
  readOnly?: string;
};

const classKeys: string[] = [
  "root",
  "disabled",
  "invalid",
  "resizable",
  "inputRoot",
  "inputRootFocused",
  "inputRootDisabled",
  "inputRootMultiline",
  "inputRootInvalid",
  "inputRootReadOnly",
  "input",
  "inputDisabled",
  "inputResizable",
  "inputReadOnly",
  "inputBorderContainer",
  "readOnly",
];

const baseInputClasses = getClasses<HvBaseInputClasses>(
  classKeys,
  "HvBaseInput"
);

export default baseInputClasses;
