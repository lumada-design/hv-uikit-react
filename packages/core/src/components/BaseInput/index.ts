export { default as baseInputClasses } from "./baseInputClasses";
export * from "./baseInputClasses";
export * from "./BaseInput";
