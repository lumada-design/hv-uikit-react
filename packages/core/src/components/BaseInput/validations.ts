import validationStates from "../Forms/FormElement/validationStates";

/**
 * Checks if the value is a number.
 *
 * @param {Number || String} num - The value to test.
 *
 * @returns {Boolean} - `true` if the value is a number `false` otherwise.
 */
const isNumeric = (num) =>
  // to prevent Number( <spaces> ) = 0
  num.trim().length > 0 && !Number.isNaN(Number(num));

/**
 * Checks if the value is an email
 *
 * @param {String} email - The value to test.
 *
 * @returns {Boolean} - `true` if the value is an email `false` otherwise.
 */
const isEmail = (email) => {
  const regexp =
    /^[^\\s]+[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?[.])+[a-z0-9](?:[a-z0-9-]*[a-z0-9])$/i;
  return regexp.test(email);
};

export const validationTypes = Object.freeze({
  none: "none",
  number: "number",
  email: "email",
});

/**
 * Returns the input's validation type based in the type of the input.
 *
 * @param {String} type - the input type.
 *
 * @returns {String}    - the validation type.
 */
export const computeValidationType = (type) => {
  switch (type) {
    case "number":
      return validationTypes.number;
    case "email":
      return validationTypes.email;

    default:
      return validationTypes.none;
  }
};

/**
 * Checks whether any integrated validation, native or not, is active.
 */
export const hasBuiltInValidations = (
  required,
  validationType,
  minCharQuantity,
  maxCharQuantity,
  validation,
  inputProps
) =>
  required ||
  validationType !== validationTypes.none ||
  minCharQuantity != null ||
  maxCharQuantity != null ||
  validation != null ||
  inputProps?.required != null ||
  inputProps?.minLength != null ||
  inputProps?.maxLength != null ||
  inputProps?.min != null ||
  inputProps?.max != null ||
  (inputProps?.type != null &&
    inputProps?.type !== "text" &&
    inputProps?.type !== "password") ||
  inputProps?.pattern != null;

/**
 * Returns the form element's validation state based in the validity state of the input.
 *
 * @param {Object} inputValidity    - the input validity state (implementing ValidityState interface).
 *
 * @returns {String}                - the validation state.
 */
export const computeValidationState = (inputValidity, isEmptyValue) => {
  // to keep 2.x behaviour,
  // consider that if the value is empty (and not required) we're returning to the standBy state.
  // might not make sense, as it makes impossible to say if the user explicitly cleared the input.
  if (inputValidity.valid && isEmptyValue) {
    return validationStates.standBy;
  }

  return inputValidity.valid
    ? validationStates.valid
    : validationStates.invalid;
};

/**
 * Returns a error message based in the validity state of the input.
 *
 * Only one error message is returned even if multiple validations fail.
 * Also, only required, minCharQuantity, maxCharQuantity and validationType have specific error messages.
 *
 * For further customization both status and statusMessage should be controlled and
 * set using the onBlur callback that receives both the value and the input validity object.
 *
 * @param {Object} inputValidity    - the input validity state (implementing ValidityState interface).
 * @param {Object} errorMessages    - the available localized error messages.
 *
 * @returns {String}                - the error message.
 */
export const computeValidationMessage = (inputValidity, errorMessages) => {
  if (inputValidity.valid) {
    return "";
  }

  if (inputValidity.valueMissing) {
    return errorMessages.requiredError;
  }
  if (inputValidity.tooLong) {
    return errorMessages.maxCharError;
  }
  if (inputValidity.tooShort) {
    return errorMessages.minCharError;
  }
  if (inputValidity.typeMismatch) {
    return errorMessages.typeMismatchError;
  }

  return errorMessages.error;
};

/**
 * Returns a object describing the validity state of the input.
 *
 * It implements the native browser's ValidityState interface:
 * https://developer.mozilla.org/en-US/docs/Web/API/ValidityState
 *
 * @param {DOMElement} input        - the input being validated.
 * @param {String} value            - the inputted value.
 * @param {Boolean} required        - if the input needs to be filled.
 * @param {Number} minCharQuantity  - the minimum length of the value.
 * @param {Number} maxCharQuantity  - the maximum length of the value.
 * @param {String} validationType   - the input value type.
 * @param {Function} validation     - a custom validation function.
 *
 * @returns {Object}                - the validity state of the input.
 */
export const validateInput = (
  input,
  value,
  required,
  minCharQuantity,
  maxCharQuantity,
  validationType,
  validation
): HvInputValidity => {
  // bootstrap validity object using browser's built-in validation
  const inputValidity: HvInputValidity = {
    valid: input?.validity?.valid ?? true,
    badInput: input?.validity?.badInput,
    customError: input?.validity?.customError,
    patternMismatch: input?.validity?.patternMismatch,
    rangeOverflow: input?.validity?.rangeOverflow,
    rangeUnderflow: input?.validity?.rangeUnderflow,
    stepMismatch: input?.validity?.stepMismatch,
    tooLong: input?.validity?.tooLong,
    tooShort: input?.validity?.tooShort,
    typeMismatch: input?.validity?.typeMismatch,
    valueMissing: input?.validity?.valueMissing,
  };

  if (!value) {
    if (required) {
      // this is redundant because we're setting the required property in the native input
      inputValidity.valueMissing = true;
      inputValidity.valid = false;
    }
  } else {
    // we're not setting minlength and maxlength in the native input
    // to avoid different browser's behaviours and keep the user experience consistent
    if (minCharQuantity !== null && value.length < minCharQuantity) {
      inputValidity.tooShort = true;
      inputValidity.valid = false;
    }

    if (maxCharQuantity !== null && value.length > maxCharQuantity) {
      inputValidity.tooLong = true;
      inputValidity.valid = false;
    }

    // the validationType is used instead of type
    // for the same reason stated above
    switch (validationType) {
      case validationTypes.number:
        if (!isNumeric(value)) {
          inputValidity.typeMismatch = true;
          inputValidity.valid = false;
        }
        break;

      case validationTypes.email:
        if (!isEmail(value)) {
          inputValidity.typeMismatch = true;
          inputValidity.valid = false;
        }
        break;

      default:
    }

    if (validation != null && !validation(value)) {
      inputValidity.customError = true;
      inputValidity.valid = false;
    }
  }

  return inputValidity;
};

export type HvInputValidity = {
  valid?: boolean;
  badInput?: boolean;
  customError?: boolean;
  patternMismatch?: boolean;
  rangeOverflow?: boolean;
  rangeUnderflow?: boolean;
  stepMismatch?: boolean;
  tooLong?: boolean;
  tooShort?: boolean;
  typeMismatch?: boolean;
  valueMissing?: boolean;
};

export const DEFAULT_ERROR_MESSAGES = {
  error: "Invalid value",
  requiredError: "The value is required",
  minCharError: "The value is too short",
  maxCharError: "The value is too long",
  typeMismatchError: "Invalid value",
};
