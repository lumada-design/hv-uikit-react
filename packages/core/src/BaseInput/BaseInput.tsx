import { forwardRef, useContext } from "react";
import MuiInputBase, { type InputBaseProps } from "@mui/material/InputBase";
import { useForkRef } from "@mui/material/utils";
import {
  useDefaultProps,
  type ExtractNames,
} from "@hitachivantara/uikit-react-utils";

import {
  buildAriaPropsFromContext,
  buildFormElementPropsFromContext,
  HvFormElementContext,
  HvFormElementDescriptorsContext,
} from "../FormElement";
import { staticClasses, useClasses } from "./BaseInput.styles";

export { staticClasses as baseInputClasses };

export type HvBaseInputClasses = ExtractNames<typeof useClasses>;

export interface HvBaseInputProps
  extends Omit<
    InputBaseProps,
    "onChange" | "classes" | "ref" | "color" | "size" | "inputProps"
  > {
  /** The input name. */
  name?: string;
  /** The value of the input, when controlled. */
  value?: React.InputHTMLAttributes<HTMLInputElement>["value"];
  /** The initial value of the input, when uncontrolled. */
  defaultValue?: React.InputHTMLAttributes<HTMLInputElement>["value"];
  /** If `true` the input is disabled. */
  disabled?: boolean;
  /** Indicates that the input is not editable. */
  readOnly?: boolean;
  /** If true, the input element will be required. */
  required?: boolean;
  /**
   * Callback fired when the value is changed.
   *
   * You can pull out the new value by accessing `event.target.value` (string),
   * or using the second callback argument.
   */
  onChange?: (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
    value: string,
  ) => void;
  /** If true, a textarea element will be rendered. */
  multiline?: boolean;
  /** If true and multiline is also true the textarea element will be resizable. */
  resizable?: boolean;
  /** Denotes if the input is in an invalid state. */
  invalid?: boolean;
  /**
   * Allows passing a ref to the underlying input
   * @deprecated Use `ref` directly instead
   * */
  inputRef?: InputBaseProps["inputRef"];
  /**
   * [Attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#Attributes) applied to the `input` element.
   */
  inputProps?: React.InputHTMLAttributes<
    HTMLInputElement | HTMLTextAreaElement
  >;
  /** A Jss Object used to override or extend the styles applied to the component. */
  classes?: HvBaseInputClasses;
}

/**
 * An Input component that only posses the most basic functionalities.
 * It should be used alongside the other form elements to construct a proper accessible form.
 */
export const HvBaseInput = forwardRef<
  // no-indent
  React.ElementRef<"input">,
  HvBaseInputProps
>(function HvBaseInput(props, ref) {
  const {
    classes: classesProp,
    className = "",
    id,
    name,
    value,
    defaultValue,
    required,
    readOnly,
    disabled,
    onChange,
    type = "text",
    placeholder,
    multiline,
    resizable,
    invalid: invalidProp,
    inputRef,
    inputProps = {},
    ...others
  } = useDefaultProps("HvBaseInput", props);
  const { classes, cx } = useClasses(classesProp);
  const formElementContext = useContext(HvFormElementContext);
  const formElementProps = buildFormElementPropsFromContext(
    name,
    disabled,
    readOnly,
    required,
    formElementContext,
  );

  const forkedRef = useForkRef(ref, inputRef);

  const invalid = invalidProp || formElementProps.status === "invalid";

  const formElementDescriptorsContext = useContext(
    HvFormElementDescriptorsContext,
  );
  const ariaProps = buildAriaPropsFromContext(
    inputProps,
    formElementDescriptorsContext,
    invalid,
    id,
  );

  return (
    <MuiInputBase
      id={id}
      name={formElementProps.name}
      value={value}
      defaultValue={defaultValue}
      placeholder={placeholder}
      readOnly={!!formElementProps.readOnly}
      disabled={formElementProps.disabled}
      onChange={(event) => onChange?.(event, event.target.value)}
      className={cx(classes.root, className, {
        [classes.multiline]: multiline,
        [classes.invalid]: invalid,
        [classes.readOnly]: formElementProps.readOnly,
        [classes.disabled]: formElementProps.disabled,
      })}
      classes={{
        focused: classes.focused,
        input: cx(classes.input, {
          [classes.inputResizable]: !formElementProps.disabled && resizable,
          [classes.inputDisabled]: formElementProps.disabled,
          [classes.inputReadOnly]: formElementProps.readOnly,
        }),
      }}
      inputProps={{
        // Avoid the required attribute at the root node
        required: formElementProps.required,
        ...inputProps,
        ...ariaProps,
      }}
      inputRef={forkedRef}
      multiline={multiline}
      {...(multiline ? { rows: 10 } : { type })}
      {...others}
    />
  );
});
