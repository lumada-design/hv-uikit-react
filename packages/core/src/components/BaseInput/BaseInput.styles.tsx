import styled from "@emotion/styled";
import {
  Input as MuiInput,
  inputClasses as MuiInputClasses,
} from "@mui/material";
import { theme } from "@hitachivantara/uikit-styles";
import { transientOptions } from "@core/utils/transientOptions";
import { outlineStyles } from "@core/utils";
import baseInputClasses from "./baseInputClasses";
import { CSSProperties } from "react";

export const StyledRoot = styled(
  "div",
  transientOptions
)(
  ({
    $disabled,
    $invalid,
    $resizable,
    $readOnly,
  }: {
    $disabled: boolean;
    $invalid: boolean;
    $resizable: boolean;
    $readOnly: boolean;
  }) => ({
    display: "inline-block",
    width: "100%",
    position: "relative",

    [`&:hover .${baseInputClasses.inputBorderContainer}`]: {
      backgroundColor: theme.baseInput.hoverColor,
    },

    [`&:focus-within .${baseInputClasses.inputBorderContainer}`]: {
      backgroundColor: theme.baseInput.hoverColor,
    },

    ...($resizable && {
      width: "auto",
    }),

    ...($readOnly && {
      [`& .${baseInputClasses.inputBorderContainer}`]: {
        backgroundColor: "transparent",
      },

      [`&:hover .${baseInputClasses.inputBorderContainer}`]: {
        backgroundColor: "transparent",
      },

      [`&:focus-within .${baseInputClasses.inputBorderContainer}`]: {
        backgroundColor: "transparent",
      },
    }),

    ...($disabled && {
      [`& .${baseInputClasses.inputBorderContainer}`]: {
        backgroundColor: theme.colors.atmo4,
      },

      [`&:hover .${baseInputClasses.inputBorderContainer}`]: {
        backgroundColor: theme.colors.atmo4,
      },
    }),

    ...($invalid && {
      "&:not(.disabled)": {
        [`& .${baseInputClasses.inputBorderContainer}`]: {
          backgroundColor: theme.colors.negative,
        },

        [`&:hover .${baseInputClasses.inputBorderContainer}`]: {
          backgroundColor: theme.colors.negative,
        },
      },
    }),
  })
);

export const StyledInputBorderContainer = styled("div")({
  position: "absolute",
  width: "calc(100% - 4px)",
  height: theme.baseInput.underlineHeight,
  top: "31px",
  left: "2px",
  backgroundColor: theme.colors.atmo4,
});

export const StyledInput = styled(
  MuiInput,
  transientOptions
)(
  ({
    $resizable,
    $disabled,
    $readOnly,
    $invalid,
  }: {
    $resizable: boolean;
    $disabled: boolean;
    $readOnly: boolean;
    $invalid: boolean;
  }) => ({
    [`&.${MuiInputClasses.root}`]: {
      margin: 0,
      width: "100%",
      borderRadius: theme.radii.base,
      height: "32px",
      border: `1px solid ${theme.baseInput.borderColor}`,
      boxSizing: "border-box",
      backgroundColor: theme.colors.atmo1,
      fontFamily: theme.fontFamily.body,

      ...($invalid && {
        borderColor: theme.colors.negative,
      }),

      ...($readOnly && {
        borderColor: theme.baseInput.readOnlyBorderColor,
        backgroundColor: theme.baseInput.readOnlyBackgroundColor,
      }),

      [`&:hover:not(.${baseInputClasses.inputRootDisabled}):not(.${baseInputClasses.inputRootInvalid}):not(.${baseInputClasses.inputRootReadOnly})`]:
        {
          borderColor: theme.baseInput.hoverColor,
        },

      [`&:hover:not(.${baseInputClasses.inputRootDisabled})::before`]: {
        borderBottom: "none",
      },

      "&::before": {
        borderBottom: "none",
      },

      "&::after": {
        borderBottom: "none",
      },
    },

    [`& .${MuiInputClasses.input}`]: {
      marginLeft: theme.space.xs,
      marginRight: theme.space.xs,
      padding: "6px 0 5px",
      overflow: "hidden",
      textOverflow: "ellipsis",
      outline: "none",
      width: "initial",
      flexGrow: 1,
      ...(theme.typography.body as CSSProperties),

      "&::placeholder": {
        opacity: 1,
        color: theme.baseInput.placeholderColor,
      },

      "&::-ms-clear": {
        display: "none",
      },

      ...($resizable && {
        resize: "both",
        width: "100%",
      }),

      ...($disabled && {
        color: theme.baseInput.disabledTextColor,
      }),

      ...($readOnly && {
        color: theme.baseInput.readOnlyTextColor,
      }),
    },

    [`&.${MuiInputClasses.focused}`]: {
      background: theme.colors.atmo1,
      ...outlineStyles,

      "&:hover": {
        background: theme.colors.atmo1,
      },

      [`&.${baseInputClasses.inputRootReadOnly}`]: {
        backgroundColor: theme.baseInput.readOnlyBackgroundColor,
      },
    },

    [`&.${MuiInputClasses.disabled}`]: {
      background: theme.baseInput.disabledBackgroundColor,
      borderColor: theme.baseInput.disabledBorderColor,
      cursor: "not-allowed",

      "&:hover": {
        background: theme.baseInput.disabledBackgroundColor,
        cursor: "not-allowed",
      },

      "&&::before": {
        borderBottomStyle: "none",
      },
    },

    [`&.${MuiInputClasses.multiline}`]: {
      padding: 0,
      backgroundColor: "transparent",
      overflow: "auto",
      border: "none",
      height: "auto",

      [`& .${baseInputClasses.input}`]: {
        border: `1px solid ${theme.baseInput.multilineBorderColor}`,
        borderRadius: theme.radii.base,
        backgroundColor: theme.colors.atmo1,
        height: "auto",
        minHeight: "21px",
        padding: "5px 10px",
        overflow: "auto",
        marginLeft: "0px",
        marginRight: "0px",

        "&:hover": {
          border: `1px solid ${theme.baseInput.hoverColor}`,
        },

        ...($disabled && {
          backgroundColor: theme.baseInput.disabledBackgroundColor,
          border: `1px solid ${theme.baseInput.multilineDisabledBorderColor}`,

          ":hover": {
            backgroundColor: theme.baseInput.disabledBackgroundColor,
            border: `1px solid ${theme.baseInput.multilineDisabledBorderColor}`,
          },
        }),

        ...($readOnly && {
          border: `1px solid ${theme.baseInput.readOnlyBorderColor}`,
          backgroundColor: theme.baseInput.readOnlyBackgroundColor,

          "&:hover": {
            border: `1px solid ${theme.baseInput.readOnlyBorderColor}`,
            backgroundColor: theme.baseInput.readOnlyBackgroundColor,
          },

          "&:focus-within": {
            border: `1px solid ${theme.baseInput.readOnlyBorderColor}`,
            backgroundColor: theme.baseInput.readOnlyBackgroundColor,
          },
        }),

        ...($invalid && {
          border: `1px solid ${theme.colors.negative}`,

          "&:hover": {
            border: `1px solid ${theme.colors.negative}`,
          },

          "&:focus-within": {
            border: `1px solid ${theme.colors.negative}`,
          },
        }),
      },
    },
  })
);
