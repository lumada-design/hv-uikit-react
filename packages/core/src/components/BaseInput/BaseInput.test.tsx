import { fireEvent, render, screen } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";
import { HvBaseInput } from "./BaseInput";

describe("BaseInput", () => {
  it("should mark the input as disabled", () => {
    render(<HvBaseInput disabled />);
    expect(screen.getByRole("textbox")).toBeDisabled();
  });

  it("should mark the input as required", () => {
    render(<HvBaseInput required />);
    expect(screen.getByRole("textbox")).toBeRequired();
  });

  it("should mark the input as read only", () => {
    render(<HvBaseInput readOnly />);
    expect(screen.getByRole("textbox")).toHaveAttribute("readOnly");
  });

  it("should render a textarea in multiline mode", () => {
    render(<HvBaseInput multiline />);
    expect(screen.getByRole("textbox")).toBeInTheDocument();
  });

  it("should call the onChange callback", () => {
    const mockFn = vi.fn();
    render(<HvBaseInput onChange={mockFn} />);
    const input = screen.getByRole("textbox");
    fireEvent.change(input, { target: { value: "mockText" } });
    expect(mockFn).toHaveBeenCalled();
  });
});
