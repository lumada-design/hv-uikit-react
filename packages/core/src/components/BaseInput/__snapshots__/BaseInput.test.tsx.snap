// Vitest Snapshot v1

exports[`BaseInput > should render correctly 1`] = `
<div>
  <div
    class="HvBaseInput-root css-fp1w37-StyledRoot e8ibl7d2"
  >
    <div
      class="MuiInputBase-root MuiInput-root HvBaseInput-inputRoot MuiInput-underline MuiInputBase-colorPrimary  e8ibl7d0 css-vw6roc-MuiInputBase-root-MuiInput-root-StyledInput"
    >
      <input
        class="MuiInputBase-input MuiInput-input HvBaseInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
        rows="10"
        type="text"
        value=""
      />
    </div>
    <div
      class="HvBaseInput-inputBorderContainer css-u8jwkr-StyledInputBorderContainer e8ibl7d1"
      role="presentation"
    />
  </div>
</div>
`;
